### PR TITLE
Feature/univar polynomial commit

### PR DESCRIPTION
**Pull request by nooma-42**
based on baloo univariate sumcheck.

### CODE REVIEW

File: src/common_util/univariatesumcheck.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    """
+    
+    # FIXME: ftt mind need to rotate as well
+    # long division to get XR(X) and Q2(X)
+    
+    #phi_alpha = phi_x.barycentric_eval(alpha)
+    phi_alpha = phi_x.coeff_eval(alpha) # FIXME: root of unity might need to be rotated
+    Q2_x, XR_x = ((D_x * t_x - phi_alpha)).div_with_remainder(zI_x)    
+    x = Polynomial(list(map(Scalar, [0, 1])), Basis.MONOMIAL)
+    R_x = XR_x / x
```
Comment (luckyyang) [inline src/common_util/univariatesumcheck.py]: The computation for `Q2_x, XR_x` seems not right, you could refer to my code at line `#271`  in file https://github.com/NOOMA-42/pylookup/pull/5/files#diff-6458f24a7ddd0cfb4aa1bf4a9c04854b086c40e2ed309c341d968f6ebba3051dR271

File: src/common_util/poly.py [hunk 2]
```diff
@@ -144,12 +144,48 @@ def __truediv__(self, other):
...
+        elif exponent == 1:
+            return self
+        else:
+            # Repeated multiplication for exponentiation
+            result = self
+            for _ in range(exponent - 1):
+                result = result * self
+            return result
+
+    def div_with_remainder(self, other):
```
Comment (luckyyang) [inline src/common_util/poly.py:170]: This method shouldn't be used, all polynomial division we used should not have reminder

File: src/common_util/univariatesumcheck.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    """ 
+    Input:
+        M = [
+            [0, 1],
+            [1, 0],
+            [0, 1],
+        ]
+    Returns:
+        col: [m] → [k] is such that col(j) = i if and only if Mj,i is the only non-zero element in row j of M (P.8 in Baloo)
+        [2, 1, 2]
```
Comment (luckyyang) [inline src/common_util/univariatesumcheck.py:32]: Should be [1, 0, 1]?

File: src/common_util/univariatesumcheck_test.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        col_mapping = create_col_mapping(M)
+        assert list(col_mapping) == [3, 2, 3]
+    
+    def test_construct_Dx(self) -> None:        
+        m = len(M) # number of rows = number of mu
+        N = len(c) # table size
+
+        V = Scalar.roots_of_unity(m) # set for mu
+        H = Scalar.roots_of_unity(N) # set for rho
+        H = left_rotate_root_of_unity(H)
```
Comment (luckyyang) [inline src/common_util/univariatesumcheck_test.py]: Do not have to rotate, you can just use 1 as the first item of SRS: 1, w, w^2...

**Review by luckyyang — CHANGES_REQUESTED** [review]

File: src/common_util/univariatesumcheck.py (hunk 1 above)
Comment (nooma-42) [reply src/common_util/univariatesumcheck.py]: If we treat D_x * t_x as dividend and z_I as divisor, isn't XR_X the residue?
截圖 2024-02-23 下午11 01 31

ref: P.16

Comment (NOOMA-42) [reply src/common_util/univariatesumcheck.py:142]: ah I see, you minus residue first?

Comment (luckyyang) [reply src/common_util/univariatesumcheck.py]: No, XR_X is not the residue. You can only calculate with this way: XR_X = Σ_i(a_i * b_i * normalized_lag_poly(root)) - Σ_i(a_i * b_i)

Comment (luckyyang) [reply src/common_util/univariatesumcheck.py]: If we get residue/remainder, then there must be something wrong with our polynomial construction

Comment (luckyyang) [reply src/common_util/univariatesumcheck.py]: Will change my code with this way

File: src/common_util/univariatesumcheck.py [hunk 5]
```diff
@@ -98,46 +98,47 @@ def construct_Dx(M: list[list], V: list[Scalar], H: list[Scalar], vec_a: list, a
...
+
     #phi_alpha = phi_x.barycentric_eval(alpha)
     phi_alpha = phi_x.coeff_eval(alpha)
-    Q2_x, XR_x = ((D_x * t_x - phi_alpha)).div_with_remainder(zI_x)    
-    x = Polynomial(list(map(Scalar, [0, 1])), Basis.MONOMIAL)
-    R_x = XR_x / x
+    Q2_x, XR_x = ((D_x * t_x - phi_alpha)).div_with_remainder(zI_x)
+    # x = Polynomial(list(map(Scalar, [0, 1])), Basis.MONOMIAL)
+    # R_x = XR_x / x
+    R_x = XR_x
```
Comment (nooma-42) [inline src/common_util/univariatesumcheck.py:141]: Why can we skip division?

File: src/common_util/univariatesumcheck_test.py [hunk 6]
```diff
@@ -126,19 +126,18 @@ def test_verify(self) -> None:
         univariate_sumcheck_verify(pi, (t_x, zI_x), setup)
 
     def test_multilinear(self) -> None:
-        v = Polynomial(list(map(Scalar, [1, 2])), Basis.LAGRANGE)
+        table = [1, 2, 3, 4]
```
Comment (NOOMA-42) [inline src/common_util/univariatesumcheck_test.py:129]: I'm not sure I understand this

**Pull request closed without merging**